File: ios-app/LvtCrm/Data/Auth/CredentialStore.swift
import Foundation

struct CredentialSnapshot: Equatable, Sendable {
    let accessToken: String
    let refreshToken: String
    let revision: Int64
}

protocol CredentialStore: AnyObject, Sendable {
    var accessToken: String? { get }
    var refreshToken: String? { get }

    func snapshot() -> CredentialSnapshot?
    func invalidatePendingWrites() -> Int64
    func saveIfRevision(_ revision: Int64, accessToken: String, refreshToken: String) -> Bool
    func replaceIfCurrent(_ expected: CredentialSnapshot, accessToken: String, refreshToken: String) -> Bool
    func clearIfRevision(_ revision: Int64) -> Bool
    func clearIfCurrent(_ expected: CredentialSnapshot) -> Bool
    func clear()
}

File: android-app/app/src/main/java/lvt/crm/data/auth/CredentialStore.kt
package lvt.crm.data.auth

data class CredentialSnapshot(
    val accessToken: String,
    val refreshToken: String,
    val revision: Long,
)

interface CredentialStore {
    val accessToken: String?
    val refreshToken: String?

    fun snapshot(): CredentialSnapshot?
    fun invalidatePendingWrites(): Long
    fun saveIfRevision(revision: Long, accessToken: String, refreshToken: String): Boolean
    fun replaceIfCurrent(
        expected: CredentialSnapshot,
        accessToken: String,
        refreshToken: String,
    ): Boolean
    fun clearIfRevision(revision: Long): Boolean
    fun clearIfCurrent(expected: CredentialSnapshot): Boolean
    fun clear()
}
